libwasmvm/src/handle_vm_panic.rs

```diff
@@ -0,0 +1,34 @@
+use std::any::Any;
+
+/// A function to process cases in which the VM panics.
+///
+/// We want to provide as much debug information as possible
+/// as those cases are not expated to happen during healthy operations.
+pub fn handle_vm_panic(what: &str, err: Box<dyn Any + Send + 'static>) {
+    eprintln!("Panic in {what}:");
+    eprintln!("{err:?}"); // Does not show useful information, see https://users.rust-lang.org/t/return-value-from-catch-unwind-is-a-useless-any/89134/6
+    eprintln!(
+        "This indicates a panic in during the operations of libwasmvm/cosmwasm-vm.
+Such panics must not happen and are considered bugs. If you see this in any real-world or
+close-to-real-world usage of wasmvm, please consider filing a security report,
+no matter if it can be abused or not:
+(https://github.com/CosmWasm/advisories/blob/main/SECURITY.md#reporting-a-vulnerability).
+Thank you for your help keeping CosmWasm safe and secure 💚"
+    );
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::panic::catch_unwind;
+
+    #[test]
+    fn handle_vm_panic_works() {
+        fn nice_try() {
+            panic!("oh no!");
+        }
+        let err = catch_unwind(nice_try).unwrap_err();
+        handle_vm_panic("nice_try", err);
+    }
```

Comment:
Maybe a bit of a nitpick, but I'm not sure how I feel about this test. It's really testing `catch_unwind` rather than `handle_vm_panic`. Basically the only thing this ensures about our code is that `handle_vm_panic` doesn't panic itself.

Reply:
I agree for the automated case and it might not be a text book style test. However, it is very handy for manual inspection of the console output (which revealed the `Any` debug problem) as well as ensuring the function can be used in isolation with the weird input type. Do you have an idea how to improve the test?

Reply:
The guard pattern maybe? A bit verbose but no parameter needed?

```rust
static VM_WRITER: RefCell<Option<Box<dyn Writer>>> = RefCell::new(None);

fn with_writer<W: Write + 'static, F: Fn(&dyn Writer)>(w: W, func: F) {
    let prev = VM_WRITER.get();
    VM_WRITER.set(Some(Box::new(w)));
    access_writer(func);
    VM_WRITER.set(prev);
}

fn access_writer<F: Fn(&dyn Writer)>(func: F) {
    VM_WRITER.with_borrow(func);
}

[...]

access_writer(|w| writeln!(w, "hello world!"));
```

Reply:
Well, that code itself won't work immediately because `.with_borrow` isn't enough and it needs a fallback case to `io::stdout()` but close enough to illustrate what I mean.

Reply:
And the overhead is alright since it's a rare case to occur anyway (hopefully never)

Reply:
No strong opinion on if/how to improve this. But if there are no easy wins, I'd merge this as is and leave it up to you to improve it later.

Reply:
Alright, then let's merge as is for now.